pmd-scala/src/main/java/net/sourceforge/pmd/lang/scala/ast/ScalaTreeBuilder.java

```diff
@@ -0,0 +1,385 @@
+import net.sourceforge.pmd.lang.scala.ast.nodes.ASTTermSelect;
+import net.sourceforge.pmd.lang.scala.ast.nodes.ASTTermSuper;
+import net.sourceforge.pmd.lang.scala.ast.nodes.ASTTermThis;
+import net.sourceforge.pmd.lang.scala.ast.nodes.ASTTermThrow;
+import net.sourceforge.pmd.lang.scala.ast.nodes.ASTTermTry;
+import net.sourceforge.pmd.lang.scala.ast.nodes.ASTTermTryWithHandler;
+import net.sourceforge.pmd.lang.scala.ast.nodes.ASTTermTuple;
+import net.sourceforge.pmd.lang.scala.ast.nodes.ASTTermWhile;
+import net.sourceforge.pmd.lang.scala.ast.nodes.ASTTermXml;
+import net.sourceforge.pmd.lang.scala.ast.nodes.ASTTypeAnd;
+import net.sourceforge.pmd.lang.scala.ast.nodes.ASTTypeAnnotate;
+import net.sourceforge.pmd.lang.scala.ast.nodes.ASTTypeApply;
+import net.sourceforge.pmd.lang.scala.ast.nodes.ASTTypeApplyInfix;
+import net.sourceforge.pmd.lang.scala.ast.nodes.ASTTypeBounds;
+import net.sourceforge.pmd.lang.scala.ast.nodes.ASTTypeByName;
+import net.sourceforge.pmd.lang.scala.ast.nodes.ASTTypeExistential;
+import net.sourceforge.pmd.lang.scala.ast.nodes.ASTTypeFunction;
+import net.sourceforge.pmd.lang.scala.ast.nodes.ASTTypeImplicitFunction;
+import net.sourceforge.pmd.lang.scala.ast.nodes.ASTTypeLambda;
+import net.sourceforge.pmd.lang.scala.ast.nodes.ASTTypeMethod;
+import net.sourceforge.pmd.lang.scala.ast.nodes.ASTTypeName;
+import net.sourceforge.pmd.lang.scala.ast.nodes.ASTTypeOr;
+import net.sourceforge.pmd.lang.scala.ast.nodes.ASTTypeParam;
+import net.sourceforge.pmd.lang.scala.ast.nodes.ASTTypePlaceholder;
+import net.sourceforge.pmd.lang.scala.ast.nodes.ASTTypeProject;
+import net.sourceforge.pmd.lang.scala.ast.nodes.ASTTypeRefine;
+import net.sourceforge.pmd.lang.scala.ast.nodes.ASTTypeRepeated;
+import net.sourceforge.pmd.lang.scala.ast.nodes.ASTTypeSelect;
+import net.sourceforge.pmd.lang.scala.ast.nodes.ASTTypeSingleton;
+import net.sourceforge.pmd.lang.scala.ast.nodes.ASTTypeTuple;
+import net.sourceforge.pmd.lang.scala.ast.nodes.ASTTypeVar;
+import net.sourceforge.pmd.lang.scala.ast.nodes.ASTTypeWith;
+import net.sourceforge.pmd.lang.scala.ast.nodes.AbstractScalaNode;
+
+import scala.meta.Case;
+import scala.meta.Ctor;
+import scala.meta.Decl;
+import scala.meta.Defn;
+import scala.meta.Enumerator;
+import scala.meta.Import;
+import scala.meta.Importee;
+import scala.meta.Importer;
+import scala.meta.Init;
+import scala.meta.Lit;
+import scala.meta.Mod;
+import scala.meta.Name;
+import scala.meta.Pat;
+import scala.meta.Pkg;
+import scala.meta.Self;
+import scala.meta.Source;
+import scala.meta.Template;
+import scala.meta.Term;
+import scala.meta.Tree;
+import scala.meta.Tree.Quasi;
+import scala.meta.Type;
+
+/**
+ * Translates Scala's AST to a PMD-compatible AST.
+ *
+ */
+public class ScalaTreeBuilder {
+
+    private static final Map<Class<? extends Tree>, Constructor<? extends AbstractScalaNode<?>>> NODE_TYPE_TO_NODE_ADAPTER_TYPE = new HashMap<>();
+
+    static {
+        register(Case.class, ASTCase.class);
```

Comment:
FYI I kept the AST here to keep in line with other languages. I'm not sure if this is the right decision as this means the XPath node if Case, but the Java node is ASTCase.class. What's your take?

Reply:
This is the case in all other language implementations, the "AST" prefix is noisy in XPath, and only has some value added in Java code (eg to differentiate the PMD node wrapper from the node it wraps)

Reply:
Ok I'll keep as is then